### PR TITLE
feat(#6289): Bumps CouchDb version to 3.3.2

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

User roles are added to db security on settings change. 
Config remains unchanged with a couple of exceptions: 
- `request_timeout` no longer `infinity` because of [this error](https://github.com/apache/couchdb/issues/3789)
- added `[attachments]` section back, as cht-conf depends on those settings to exist and it isn't part of default config anymore

medic/cht-core#6289

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7964-couch3/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7964-couch3/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:7964-couch3/docker-compose/cht-couchdb-clustered.yml)
 
# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: couchdb/10-docker-default.ini
Comment (dianabarsan) [inline couchdb/10-docker-default.ini:4]: When using `infinity`, I got an error similar to this one: https://github.com/apache/couchdb/issues/3789
```
[error] 2023-02-27T12:10:48.341702Z couchdb@couchdb-1.local <0.754.0> 08bd1a6cec req_err(2329098955) unknown_error : timeout_value
    [<<"fabric_util:get_shard/4 L149">>,<<"fabric:get_security/2 L220">>,<<"chttpd_auth_request:db_authorization_check/1 L109">>,<<"chttpd_auth_request:authorize_request/1 L19">>,<<"chttpd:handle_req_after_auth/2 L406">>,<<"chttpd:process_request/1 L388">>,<<"chttpd:handle_request_int/1 L325">>,<<"mochiweb_http:headers/6 L140">>]
[notice] 2023-02-27T12:10:48.341813Z couchdb@couchdb-1.local <0.754.0> 08bd1a6cec haproxy:5984 192.168.240.8 admin GET /_users 500 ok 1
```

File: api/src/db.js
Comment (garethbowen) [inline api/src/db.js:171]: I think if you call `addRoleToSecurity` multiple times this default object will mutate because it's never cloned. You could either clone this on access on line 184, or I think a better solution would be to change this constant a getter...

```suggestion
  const getDefaultSecurityStructure = () => {
    return { names: [], roles: [] };
  };
```

File: api/src/db.js
Comment (garethbowen) [inline api/src/db.js:175]: I think it'd be better to throw here to make it clear that it didn't succeed.

File: api/src/db.js
Comment (garethbowen) [inline api/src/db.js:188]: The second parameter of the Error constructor shouldn't be a random object. I suspect you switched from `console.error` to `throw new Error` at some point?

File: api/src/db.js
Comment (garethbowen) [inline api/src/db.js:187]: This is weird right, because we've found the security object in the db and now we're throwing, which will be a bad time if this is run during migration. Can't we just add the `roles` array to proactively "fix" the security object?

File: api/src/db.js
Comment (garethbowen) [inline api/src/db.js:173]: Taking `addAsAdmin` as a boolean makes it hard to read when invoking it... Consider adding two functions: `addRoleToAdmins` and `addRoleToMembers` to make it clear.

File: couchdb/10-docker-default.ini
Comment (garethbowen) [reply couchdb/10-docker-default.ini:4]: I wonder if it should be 1 hour to match our load balancer timeouts?

File: couchdb/10-docker-default.ini
Comment (garethbowen) [inline couchdb/10-docker-default.ini:13]: Why do we need this?

File: couchdb/Dockerfile
Comment (garethbowen) [inline couchdb/Dockerfile:1]:
```suggestion
FROM couchdb:3.3.2 as base_couchdb_build
```

**Review by garethbowen — CHANGES_REQUESTED** [review]
Minor suggestions inline.

Exciting stuff!

File: couchdb/10-docker-default.ini
Comment (dianabarsan) [reply couchdb/10-docker-default.ini:13]: We have this ancient migration that adds `national_admin` to `_users`: https://github.com/medic/cht-core/blob/master/api/src/migrations/add-national_admin-role.js#L58
We have historically allowed `national_admins` to make changes to users, we would be breaking backwards compatibility.

File: couchdb/10-docker-default.ini
Comment (dianabarsan) [reply couchdb/10-docker-default.ini:4]: I think that not all requests get to CouchDb through load balancer. On the other hand, I have no strong feelings about this timeout.

File: couchdb/10-docker-default.ini
Comment (garethbowen) [reply couchdb/10-docker-default.ini:13]: I think it's probably time to break that backwards compatibility. I'm worried that changing this setting puts us at a security risk, and the national_admin role has been deprecated for years.

What do you think about a migration to add the `_admin` role to all `national_admin`s?

This could be done as a separate PR.

File: couchdb/10-docker-default.ini
Comment (garethbowen) [reply couchdb/10-docker-default.ini]: I agree... I suspect this was set for things like the sentinel db watcher which could theoretically run for a year. But it's probably a bad idea for it to run for a year, and having a timeout is probably a good idea. Happy to leave this as is and discuss separately.

File: couchdb/10-docker-default.ini
Comment (dianabarsan) [reply couchdb/10-docker-default.ini:13]: Hm, ok. I'm thinking about other ways we would break functionality by making all `national_admin`s actual admins. 
This issue comes to mind: https://github.com/medic/cht-core/issues/8243 where we would have previously allowed to turn off permissions for `national_admins`.

**Review by garethbowen — APPROVED** [review]
:tada:

File: couchdb/10-docker-default.ini
Comment (kennsippell) [inline couchdb/10-docker-default.ini:30]: would be great to expose prometheus monitoring via an unauthenticated endpoint https://github.com/medic/cht-watchdog/issues/76.  happy to do this in a second PR after this merges

Comment (fardarter) [reply couchdb/10-docker-default.ini:30]: Can I suggest you develop a user configurable config layer? We're developing against this and having to modify directly means it's difficult to stay conformant with upstream. 

See for example:

```suggestion
echo "
[daemons]
httpsd = {couch_httpd, start_link, [https]}
[admins]
admin = ${COUCHDB_PASSWORD}
[ssl]
port = 6984
enable = true
cert_file = /opt/couchdb/etc/certs/cert.pem
key_file = /opt/couchdb/etc/certs/privkey.pem
cacert_file = /opt/couchdb/etc/certs/chain.pem" |

sudo crudini --merge /opt/couchdb/etc/local.ini
```

Comment (fardarter) [reply couchdb/10-docker-default.ini:30]: Can I suggest you develop a user configurable config layer? We're developing against this and having to modify directly means it's difficult to stay conformant with upstream. 

See for example:

```suggestion
echo "
[daemons]
httpsd = {couch_httpd, start_link, [https]}
[admins]
admin = ${COUCHDB_PASSWORD}
[ssl]
port = 6984
enable = true
cert_file = /opt/couchdb/etc/certs/cert.pem
key_file = /opt/couchdb/etc/certs/privkey.pem
cacert_file = /opt/couchdb/etc/certs/chain.pem" |

sudo crudini --merge /opt/couchdb/etc/local.ini
```

Comment (dianabarsan) [reply couchdb/10-docker-default.ini:30]: Hi @fardarter 

Thanks a lot for this great suggestion! 
I think this is out of scope of this PR, and this has been discussed while we originally developed 4.x. We have not reached a conclusion, and we decided this would be a goal for later. 

Would you mind creating an issue that details exactly which configs you need and how you're thinking about persisting them in the container? 
Thanks a lot!

Comment (fardarter) [reply couchdb/10-docker-default.ini:30]: Our current plan is to build our own version of the images with patches as needed, so this would be a pre-image build step. Would argue that supporting user patching behaviour would help with keeping people contributing to upstream.

I've set up a system where we're doing our best to only patch what we need. Could be improved; happy to demo it.

I'm actually trying to find your contact details at the moment (which is why I'm on this PR). Don't suppose I could give you mine (and then we could delete the comment)?

My senior dev on this project has fallen ill and he was using this PR as a way to let us use 3.3.2 (we'd like to use the bitnami image on azure, but I can script those differences).

I'm supposed to be doing ops so I'm a little lost as to whether we can even use 3.3.2 at this juncture, or whether it would be foolish. I'd prefer not having to build out a 2.2.1 on a VM where there's a preconfigured, hardened bitnami image. 

I'm also a little lost as to what the right command is to test this locally, but my tests seem to be breaking.

Comment (dianabarsan) [reply couchdb/10-docker-default.ini:30]: Open discussion is welcome and desirable on this PR or on the [CHT Forum](https://forum.communityhealthtoolkit.org/). 
If you have specific questions, or have a technical challenge, the forum would be the best place to express these, so that other community members would benefit. 
I believe there already is a thread about this PR: https://forum.communityhealthtoolkit.org/t/upgrade-of-couchdb-from-version-2-3-1-to-3-3-1/2928 , I'm not sure if this is at all related to your issue. 

Please feel free to reach out on either platform! 
I'm really keen to hear more about your desired outcome. 

Thanks!

**Comment by dianabarsan** [comment]
@garethbowen are we ready for this? 
I had to include a commit that stabilizes an e2e test (same code in this PR: https://github.com/medic/cht-core/pull/8536) which was failing consistently on this branch.

File: api/tests/mocha/db.spec.js
Comment (garethbowen) [inline api/tests/mocha/db.spec.js:6]: Isn't this already set in the env?

Comment (dianabarsan) [reply api/tests/mocha/db.spec.js]: I think it should be. Removed.

File: tests/integration/cht-conf/cht-conf-actions.spec.js
Comment (garethbowen) [inline tests/integration/cht-conf/cht-conf-actions.spec.js:54]: Is this a thing??

```suggestion
        console.error(error);
```

File: tests/integration/api/server.spec.js
Comment (dianabarsan) [reply tests/integration/api/server.spec.js]: I think this is a bad merge.

File: tests/integration/cht-conf/cht-conf-actions.spec.js
Comment (dianabarsan) [reply tests/integration/cht-conf/cht-conf-actions.spec.js]: wtb compiler. :(

**Comment by garethbowen** [comment]
@dianabarsan I can't think of any reason to hold off and nothing else to test. Let's push the button and get 4.4 out.

The main concern is going to be upgrading multi-node instances, right? We don't have a good solution for that right now, but we're not likely to have one next week either so I feel like it's better to get this done now before more installations are running multi-node solutions.

**Comment by dianabarsan** [comment]
I think my main concern is not hitting the right config for us - which is the biggest unknown here. I'm merging.

**Pull request merged**
